README.md
# Mosquitto Go Auth

Mosquitto Go Auth is an authentication and authorization plugin for the Mosquitto MQTT broker.

### Intro

This is an authentication and authorization plugin for [mosquitto](https://mosquitto.org/), a well known open source MQTT broker. It's written (almost) entirely in Go: it uses cgo to expose mosquitto's auth plugin needed functions, but internally just calls Go to get everything done. 

It is greatly inspired in [jpmens'](https://github.com/jpmens) [mosquitto-auth-plug](https://github.com/jpmens/mosquitto-auth-plug).

It was intended for use with [brocaar's](https://github.com/brocaar) [Loraserver project](https://www.loraserver.io/), and thus Files, Postgres and JWT backends were the first to be developed, but more have been added. These are the backends that this plugin implements right now:

* Files
* PostgreSQL
* JWT (with local DB or remote API)
* HTTP
* Redis
* Mysql
* SQLite3
* MongoDB
* Custom (experimental)
* gRPC

**Every backend offers user, superuser and acl checks, and include proper tests.**

Please open an issue with the `feature` or `enhancement` tag to request new backends or additions to existing ones.


### Table of contents

<!-- MarkdownTOC -->

- [Requirements](#requirements)
- [Build](#build)
- [Configuration](#configuration)
	- [General options](#general-options)
	- [Cache](#cache)
	- [Log level](#log-level)
	- [Prefixes](#prefixes)
	- [Backend options](#backend-options)
- [Files](#files)
	- [Passwords file](#passwords-file)
	- [ACL file](#acl-file)
	- [Testing Files](#testing-files)
- [PostgreSQL](#postgresql)
	- [Testing Postgres](#testing-postgres)
- [Mysql](#mysql)
	- [Testing Mysql](#testing-mysql)
- [SQLite3](#sqlite3)
	- [Testing SQLite3](#testing-sqlite3)
- [JWT](#jwt)
	- [Remote mode](#remote-mode)
	- [Local mode](#local-mode)
	- [Testing JWT](#testing-jwt)
- [HTTP](#http)
	- [Response mode](#response-mode)
	- [Params mode](#params-mode)
	- [Testing HTTP](#testing-http)
- [Redis](#redis)
	- [Testing Redis](#testing-redis)
- [MongoDB](#mongodb)
	- [Testing MongoDB](#testing-mongodb)
- [Custom \(experimental\)](#custom-experimental)
	- [Testing Custom](#testing-custom)
- [gRPC](#grpc)
	- [Service](#service)
	- [Testing gRPC](#testing-grpc)
- [Benchmarks](#benchmarks)
- [Using with LoRa Server](#using-with-lora-server)
- [Docker](#docker)
- [License](#license)

<!-- /MarkdownTOC -->



### Requirements

Starting with Go 1.12 this plugin supports `Go modules` to manage dependencies. If you have `go mod` enabled, **you don't need to run any prior commands to get your dependencies.**

If you are using an older version of Go (tested with Go 1.10.3, 1.10.8 and 1.11.5) dependencies may be managed with `dep` (you may install it with `make dev-requirements` if you don't have it already **and** you've exported `$HOME/go/src/bin` to your `PATH`). Run this to have `dep` install the dependencies:

```
make requirements
```

As it interacts with mosquitto, it makes use of Cgo. Also, it (optionally) uses Redis for cache purposes.


### Build

Before building, you need to build mosquitto. For completeness, we'll build it with websockets, ssl and srv support.

First, install dependencies (tested on Debian 9):

`sudo apt-get install libwebsockets8  libwebsockets-dev libc-ares2 libc-ares-dev openssl uuid uuid-dev`

Download mosquitto and extract it (**change versions accordingly**):

```
wget http://mosquitto.org/files/source/mosquitto-1.6.3.tar.gz
tar xzvf mosquitto-1.6.3.tar.gz
cd mosquitto-1.6.3
```

Modify config.mk, setting websockets support. Then build mosquitto, add a mosquitto user and set ownership for /var/log/mosquitto and /var/lib/mosquitto/ (default log and persistence locations).

```
make
sudo make install
sudo groupadd mosquitto
sudo useradd -s /sbin/nologin mosquitto -g mosquitto -d /var/lib/mosquitto
sudo mkdir -p /var/log/mosquitto/ /var/lib/mosquitto/
sudo chown -R mosquitto:mosquitto /var/log/mosquitto/
sudo chown -R mosquitto:mosquitto /var/lib/mosquitto/
```

Finally, you may create a service for mosquitto. Create the file /etc/systemd/system/mosquitto.service with these annotations:

```
[Unit]
Description=Mosquitto MQTT v3.1/v3.1.1 server
Wants=network.target
Documentation=http://mosquitto.org/documentation/

[Service]
Type=simple
User=mosquitto
Group=mosquitto
ExecStart=/usr/local/sbin/mosquitto -c /etc/mosquitto/mosquitto.conf
Restart=on-failure
SyslogIdentifier=Mosquitto

[Install]
WantedBy=multi-user.target
```

If you are running another distro or need more details on building mosquitto, please check the offical mosquitto docs.

#### Build the plugin for mosquitto 1.4.x

Now that mosquitto is installed, building the project is fairly simple given that you meet the requirements. Just run this commands to generate go-auth.h and then go-auth.so:

```
go build -buildmode=c-archive go-auth.go
go build -buildmode=c-shared -o go-auth.so
```

or simply:

```
make
```

You can also run all tests (see Testing X for each backend's testing requirements) like this:

```
make test
```

#### Build the plugin for mosquitto 1.5.x and 1.6.x

For the latest versions of mosquitto we need to export some flags before building and then run the same commands (we'll just use make):

##### Debian (and maybe others)
```
export CGO_CFLAGS="-I/usr/local/include -fPIC"
export CGO_LDFLAGS="-shared"
make
```

This assumes that `mosquitto.h`, `mosquitto_plugin.h` and `mosquitto_broker.h` are located at `/usr/local/include`, which is true for a manually built mosquitto version in debian based systems (and probably others too).

##### MacOS
```
export CGO_CFLAGS="-I/usr/local/include -fPIC"
export CGO_LDFLAGS="-undefined dynamic_lookup"
export CGO_LDFLAGS="-shared"
make
```

This assumes that `mosquitto.h`, `mosquitto_plugin.h` and `mosquitto_broker.h` are located at `/usr/local/include`, which is true for a manually built or using brew (```brew install mosquitto```) to install mosquitto version in MacOS.

#### Raspberry Pi

To build on a Raspberry Pi (tested with Pi 3 B), you'll need to have Go installed first. You can install latest version (**last tested was 1.10.1, change it to suit your needs**) with something like this:

```
wget https://storage.googleapis.com/golang/go1.10.1.linux-armv6l.tar.gz
sudo tar -C /usr/local -xzf go1.10.1.linux-armv6l.tar.gz
```

Add Go to your path at .profile:

`
export PATH=$PATH:/usr/local/go/bin:~/go/bin 
`

Source the file (`source ~/.profile`) and check Go was correctly installed (`go version`).

Now get requirements and build as usual (just have some more patience).

##### Openssl and websockets notes

There seems to be missing packages in some Raspbian versions, so you should try to apt update before installing dependencies. Alternatively, you cand build openssl like this:

```
git clone git://git.openssl.org/openssl.git
cd openssl
./config
make
make test
sudo make install
```

For websockets support, you'll have to build libwebsockets, which needs cmake. So something like this should do the trick:

```
sudo apt-get install cmake
git clone https://github.com/warmcat/libwebsockets.git
cd libwebsockets
mkdir build
cd build
cmake ..
make
make install
```


### Configuration

The plugin is configured in [Mosquitto's](https://mosquitto.org/) configuration file (typically `mosquitto.conf`),
and it is loaded into Mosquitto auth with the ```auth_plugin``` option.


#### General options

Set path to plugin and include conf.d dir for further configuration:

```
auth_plugin /path/to/auth-plug.so
include_dir /etc/mosquitto/conf.d
```

Create some conf file (e.g., mosquitto-go-auth.conf) at /etc/mosquitto/conf.d/ and register the desired backends with:

```
auth_opt_backends files, postgres, jwt
```

#### Cache

Set cache option to true to use redis cache (defaults to false when missing). Also, set cache_reset to flush the redis DB on mosquitto startup:

```
auth_opt_cache true
auth_opt_cache_reset true
```

If `cache_reset` is set to false or omitted, cache won't be flushed upon service start.

Redis will use the following defaults if no values are given. Also, these are the available options for cache:

```
auth_opt_cache_host localhost
auth_opt_cache_port 6379
auth_opt_cache_password pwd
auth_opt_cache_db 3
auth_opt_auth_cache_seconds 30
auth_opt_acl_cache_seconds 30
```

#### Logging

You can set the log level with the `log_level` option. Valid values are: debug, info, warn, error, fatal and panic. If not set, default value is `info`.

```
auth_opt_log_level debug
```

Log destination may be set with `log_dest` option. Valid values are `stderr` (default), `stdout` and `file`. In the latter case the `log_file` option needs to be set, e.g.:

```
auth_opt_log_dest file
auth_opt_log_file /var/log/mosquitto/mosquitto.log
```

If `log_dest` or `log_file` are invalid, or if there's an error opening the file (e.g. no permissions), logging will default to `stderr`.

#### Prefixes

Though the plugin may have multiple backends enabled, there's a way to specify which backend must be used for a given user: prefixes. When enabled, `prefixes` allows to check if the username contains a predefined prefix in the form prefix_username and use the configured backend for that prefix. Options to enable and set prefixes are the following:

```
auth_opt_check_prefix true
auth_opt_prefixes filesprefix, pgprefix, jwtprefix
```

Prefixes must meet the declared backends order and number. If amounts don't match, the plugin will default to prefixes disabled.

Underscores (\_) are not allowed in the prefixes, as a username's prefix will be checked against the first underscore's index. Of course, if a username has no underscore or valid prefix, it'll be checked against all backends.


#### Backend options


debian/changelog
mosquitto-go-auth (0.6.1-1) unstable; urgency=medium

  * Debian release. Fix for compiling auth-plugin.c

 -- Myna-Project.org Srl <info@myna-project.org>  Fri, 6 Dec 2019 13:45:00 +0100

mosquitto-go-auth (0.6.1) unstable; urgency=medium

  * Fix plugin ACL checks. This release fixes ACL checks for the plugin backend.

 -- Ignacio Gómez <iegomez@uc.cl>  Tue, 1 Oct 2019 14:09:00 +0200

mosquitto-go-auth (0.6.0) unstable; urgency=medium

  * Redis MOSQL_ACL_SUBSCRIBE support. This release adds MOSQL_ACL_SUBSCRIBE support to the Redis backend, removes the default username and password for MongoDB tests and cleans up access ints at tests acl checks.

 -- Ignacio Gómez <iegomez@uc.cl>  Fri, 30 Aug 2019 21:35:00 +0200

mosquitto-go-auth (0.5.0) unstable; urgency=medium

  * gRPC support. Added gRPC backend.

 -- Ignacio Gómez <iegomez@uc.cl>  Fri, 30 Aug 2019 21:35:00 +0200

mosquitto-go-auth (0.4.0) unstable; urgency=medium

  * Migrate to MongoDB driver. The Mongo DB backend was migrated to the official driver https://github.com/mongodb/mongo-go-driver. Fixed Mongo DB auth issue. Added auth tests for Mongo DB and Redis. 

 -- Ignacio Gómez <iegomez@uc.cl>  Fri, 7 Jun 2019 17:17:00 +0200

mosquitto-go-auth (0.3.0) unstable; urgency=medium

  * Log options. Options to log to stderr, stdout and a file are now available.

 -- Ignacio Gómez <iegomez@uc.cl>  Fri, 17 May 2019 17:58:00 +0200

mosquitto-go-auth (0.2.0) unstable; urgency=medium

  * Go modules. This release adds Go modules support.

 -- Ignacio Gómez <iegomez@uc.cl>  Thu, 7 Mar 2019 18:59:00 +0100

mosquitto-go-auth (0.1.1) unstable; urgency=medium

  * Mysql and file patterns fix. Mysql now supports a unix socket connection properly. File patterns had a typo on the check that was introduced when adding mosquitto 1.5.x support. It's now been fixed.

 -- Ignacio Gómez <iegomez@uc.cl>  Mon, 25 Feb 2019 18:41:00 +0100

mosquitto-go-auth (0.1.0) unstable; urgency=medium

  * 0.1.0 mosquitto 1.5.X support. This release stabilizes support for mosquitto's latest versions. Check the readme for instructions on building for 1.5.X and 1.4.X.

 -- Ignacio Gómez <iegomez@uc.cl>  Wed, 12 Dec 2018 18:29:00 +0100



go-auth.go
package main

import "C"

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	log "github.com/sirupsen/logrus"

	b64 "encoding/base64"

	"plugin"

	goredis "github.com/go-redis/redis"
	bes "github.com/iegomez/mosquitto-go-auth/backends"
)

type Backend interface {
	GetUser(username, password string) bool
	GetSuperuser(username string) bool
	CheckAcl(username, topic, clientId string, acc int32) bool
	GetName() string
	Halt()
}

type CommonData struct {
	Backends         map[string]Backend
	Plugin           *plugin.Plugin
	PInit            func(map[string]string, log.Level) error
	PGetName         func() string
	PGetUser         func(username, password string) bool
	PGetSuperuser    func(username string) bool
	PCheckAcl        func(username, topic, clientid string, acc int) bool
	PHalt            func()
	Superusers       []string
	AclCacheSeconds  int64
	AuthCacheSeconds int64
	UseCache         bool
	RedisCache       *goredis.Client
	CheckPrefix      bool
	Prefixes         map[string]string
	LogLevel         log.Level
	LogDest          string
	LogFile          string
}

//Cache stores necessary values for Redis cache
type Cache struct {
	Host     string
	Port     string
	Password string
	DB       int32
}

var allowedBackends = map[string]bool{
	"postgres": true,
	"jwt":      true,
	"redis":    true,
	"http":     true,
	"files":    true,
	"mysql":    true,
	"sqlite":   true,
	"mongo":    true,
	"plugin":   true,
	"grpc":     true,
}

var backends []string          //List of selected backends.
var authOpts map[string]string //Options passed by mosquitto.
var cache Cache                //Cache conf.
var commonData CommonData      //General struct with options and conf.

//export AuthPluginInit
func AuthPluginInit(keys []string, values []string, authOptsNum int) {

	//Initialize Cache with default values
	cache = Cache{
		Host:     "localhost",
		Port:     "6379",
		Password: "",
		DB:       3,
	}

	log.SetFormatter(&log.TextFormatter{
		FullTimestamp: true,
	})

	superusers := make([]string, 10, 10)

	cmbackends := make(map[string]Backend)

	//Initialize common struct with default and given values
	commonData = CommonData{
		Superusers:       superusers,
		AclCacheSeconds:  30,
		AuthCacheSeconds: 30,
		CheckPrefix:      false,
		Prefixes:         make(map[string]string),
		LogLevel:         log.InfoLevel,
	}

	//First, get backends
	backendsOk := false
	authOpts = make(map[string]string)
	for i := 0; i < authOptsNum; i++ {
		if keys[i] == "backends" {
			backends = strings.Split(strings.Replace(values[i], " ", "", -1), ",")
			if len(backends) > 0 {
				backendsCheck := true
				for _, backend := range backends {
					if _, ok := allowedBackends[backend]; !ok {
						backendsCheck = false
						log.Errorf("backend not allowed: %s", backend)
					}
				}
				backendsOk = backendsCheck
			}
		} else {
			authOpts[keys[i]] = values[i]
		}
	}

	//Log and end program if backends are wrong
	if !backendsOk {
		log.Fatal("\nbackends error\n")
	}

	//Check if log level is given. Set level if any valid option is given.
	if logLevel, ok := authOpts["log_level"]; ok {

		logLevel = strings.Replace(logLevel, " ", "", -1)

		switch logLevel {
		case "debug":
			commonData.LogLevel = log.DebugLevel
		case "info":
			commonData.LogLevel = log.InfoLevel
		case "warn":
			commonData.LogLevel = log.WarnLevel
		case "error":
			commonData.LogLevel = log.ErrorLevel
		case "fatal":
			commonData.LogLevel = log.FatalLevel
		case "panic":
			commonData.LogLevel = log.PanicLevel
		default:
			log.Info("log_level unkwown, using default info level")
		}

	}

	if logDest, ok := authOpts["log_dest"]; ok {
		switch logDest {
		case "stdout":
			log.SetOutput(os.Stdout)
		case "file":
			if logFile, ok := authOpts["log_file"]; ok {
				file, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
				if err == nil {
					log.SetOutput(file)
				} else {
					log.Errorf("failed to log to file, using default stderr: %s", err)
				}
			}
		default:
			log.Info("log_dest unknown, using default stderr")
		}
	}

	//Initialize backends
	for _, bename := range backends {
		var beIface Backend
		var bErr error

		if bename == "plugin" {
			plug, plErr := plugin.Open(authOpts["plugin_path"])
			if plErr != nil {
				log.Errorf("Could not init custom plugin: %s", plErr)
				commonData.Plugin = nil
			} else {
				commonData.Plugin = plug

				plInit, piErr := commonData.Plugin.Lookup("Init")

				if piErr != nil {
					log.Errorf("Couldn't find func Init in plugin: %s", plErr)
					commonData.Plugin = nil
					continue
				}

				initFunc := plInit.(func(authOpts map[string]string, logLevel log.Level) error)

				ipErr := initFunc(authOpts, commonData.LogLevel)
				if ipErr != nil {
					log.Errorf("Couldn't init plugin: %s", ipErr)
					commonData.Plugin = nil
					continue
				}

				commonData.PInit = initFunc

				plName, gErr := commonData.Plugin.Lookup("GetName")

				if gErr != nil {
					log.Errorf("Couldn't find func GetName in plugin: %s", gErr)
					commonData.Plugin = nil
					continue
				}

				nameFunc := plName.(func() string)
				commonData.PGetName = nameFunc

				plGetUser, pgErr := commonData.Plugin.Lookup("GetUser")

				if pgErr != nil {
					log.Errorf("Couldn't find func GetUser in plugin: %s", pgErr)
					commonData.Plugin = nil
					continue
				}

				getUserFunc := plGetUser.(func(username, password string) bool)
				commonData.PGetUser = getUserFunc

				if pgErr != nil {
					log.Errorf("Couldn't find func GetUser in plugin: %s", pgErr)
					commonData.Plugin = nil
					continue
				}

				plGetSuperuser, psErr := commonData.Plugin.Lookup("GetSuperuser")

				if psErr != nil {
					log.Errorf("Couldn't find func GetSuperuser in plugin: %s", psErr)
					commonData.Plugin = nil
					continue
				}

				getSuperuserFunc := plGetSuperuser.(func(username string) bool)
				commonData.PGetSuperuser = getSuperuserFunc

				plCheckAcl, pcErr := commonData.Plugin.Lookup("CheckAcl")

				if pcErr != nil {
					log.Errorf("Couldn't find func CheckAcl in plugin: %s", pcErr)
					commonData.Plugin = nil
					continue
				}

				checkAclFunc := plCheckAcl.(func(username, topic, clientid string, acc int) bool)
				commonData.PCheckAcl = checkAclFunc

				plHalt, phErr := commonData.Plugin.Lookup("Halt")

				if phErr != nil {
					log.Errorf("Couldn't find func Halt in plugin: %s", phErr)
					commonData.Plugin = nil
					continue
				}

				haltFunc := plHalt.(func())
				commonData.PHalt = haltFunc

				log.Infof("Backend registered: %s", commonData.PGetName())

			}
		} else {
			switch bename {
			case "postgres":
				beIface, bErr = bes.NewPostgres(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["postgres"] = beIface.(bes.Postgres)
				}
			case "jwt":
				beIface, bErr = bes.NewJWT(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["jwt"] = beIface.(bes.JWT)
				}
			case "files":
				beIface, bErr = bes.NewFiles(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["files"] = beIface.(bes.Files)
				}
			case "redis":
				beIface, bErr = bes.NewRedis(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["redis"] = beIface.(bes.Redis)
				}
			case "mysql":
				beIface, bErr = bes.NewMysql(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["mysql"] = beIface.(bes.Mysql)
				}
			case "http":
				beIface, bErr = bes.NewHTTP(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["http"] = beIface.(bes.HTTP)
				}
			case "sqlite":
				beIface, bErr = bes.NewSqlite(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["sqlite"] = beIface.(bes.Sqlite)
				}
			case "mongo":
				beIface, bErr = bes.NewMongo(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["mongo"] = beIface.(bes.Mongo)
				}
			case "grpc":
				beIface, bErr = bes.NewGRPC(authOpts, commonData.LogLevel)
				if bErr != nil {
					log.Fatalf("Backend register error: couldn't initialize %s backend with error %s.", bename, bErr)
				} else {
					log.Infof("Backend registered: %s", beIface.GetName())
					cmbackends["grpc"] = beIface.(bes.GRPC)
				}
			}
		}

	}

	if cache, ok := authOpts["cache"]; ok && strings.Replace(cache, " ", "", -1) == "true" {
		log.Info("Cache activated")
		commonData.UseCache = true
	} else {
		log.Info("No cache set.")
		commonData.UseCache = false
	}

	if commonData.UseCache {
		if cacheHost, ok := authOpts["cache_host"]; ok {
			cache.Host = cacheHost
		}

		if cachePort, ok := authOpts["cache_port"]; ok {
			cache.Port = cachePort
		}

		if cachePassword, ok := authOpts["cache_password"]; ok {
			cache.Password = cachePassword
		}

		if cacheDB, ok := authOpts["cache_db"]; ok {
			db, err := strconv.ParseInt(cacheDB, 10, 32)
			if err == nil {
				cache.DB = int32(db)
			} else {
				log.Warningf("couldn't parse cache db (err: %s), defaulting to %d", err, cache.DB)
			}
		}

		if authCacheSec, ok := authOpts["auth_cache_seconds"]; ok {
			authSec, err := strconv.ParseInt(authCacheSec, 10, 64)
			if err == nil {
				commonData.AuthCacheSeconds = authSec
			} else {
				log.Warningf("couldn't parse AuthCacheSeconds (err: %s), defaulting to %d", err, commonData.AuthCacheSeconds)
			}

		}

		if aclCacheSec, ok := authOpts["acl_cache_seconds"]; ok {
			aclSec, err := strconv.ParseInt(aclCacheSec, 10, 64)
			if err == nil {
				commonData.AclCacheSeconds = aclSec
			} else {
				log.Warningf("couldn't parse AclCacheSeconds (err: %s), defaulting to %d", err, commonData.AclCacheSeconds)
			}

		}

		addr := fmt.Sprintf("%s:%s", cache.Host, cache.Port)

		//If cache is on, try to start redis.
		goredisClient := goredis.NewClient(&goredis.Options{
			Addr:     addr,
			Password: cache.Password, // no password set
			DB:       int(cache.DB),  // use default DB
		})

		_, err := goredisClient.Ping().Result()
		if err != nil {
			log.Errorf("couldn't start Redis, defaulting to no cache. error: %s", err)
			commonData.UseCache = false
		} else {
			commonData.RedisCache = goredisClient
			log.Infof("started cache redis client on DB %d", cache.DB)
			//Check if cache must be reset
			if cacheReset, ok := authOpts["cache_reset"]; ok && cacheReset == "true" {
				commonData.RedisCache.FlushDB()
				log.Infof("flushed cache")
			}
		}

	}

	if checkPrefix, ok := authOpts["check_prefix"]; ok && strings.Replace(checkPrefix, " ", "", -1) == "true" {
		//Check that backends match prefixes.
		if prefixesStr, ok := authOpts["prefixes"]; ok {
			prefixes := strings.Split(strings.Replace(prefixesStr, " ", "", -1), ",")
			if len(prefixes) == len(backends) {
				//Set prefixes
				for i, backend := range backends {
					commonData.Prefixes[prefixes[i]] = backend
				}
				log.Infof("Prefixes enabled for backends %s with prefixes %s.", authOpts["backends"], authOpts["prefixes"])
				commonData.CheckPrefix = true
			} else {
				log.Errorf("Error: got %d backends and %d prefixes, defaulting to prefixes disabled.", len(backends), len(prefixes))
				commonData.CheckPrefix = false
			}

		} else {
			log.Warn("Error: prefixes enabled but no options given, defaulting to prefixes disabled.")
			commonData.CheckPrefix = false
		}
	} else {
		commonData.CheckPrefix = false
	}

	commonData.Backends = cmbackends

}

//export AuthUnpwdCheck
func AuthUnpwdCheck(username, password string) bool {

	authenticated := false
	var cached = false
	var granted = false
	if commonData.UseCache {
		log.Debugf("checking auth cache for %s", username)
		cached, granted = CheckAuthCache(username, password)
		if cached {
			log.Debugf("found in cache: %s", username)
			return granted
		}
	}

	//If prefixes are enabled, checkt if username has a valid prefix and use the correct backend if so.
	if commonData.CheckPrefix {
		validPrefix, bename := CheckPrefix(username)
		if validPrefix {

			if bename == "plugin" {
				authenticated = CheckPluginAuth(username, password)
			} else {

				var backend = commonData.Backends[bename]

				if backend.GetUser(username, password) {
					authenticated = true
					log.Debugf("user %s authenticated with backend %s", username, backend.GetName())
				}

			}

		} else {
			//If there's no valid prefix, check all backends.
			authenticated = CheckBackendsAuth(username, password)
			//If not authenticated, check for a present plugin
			if !authenticated {
				authenticated = CheckPluginAuth(username, password)
			}
		}
	} else {
		authenticated = CheckBackendsAuth(username, password)
		//If not authenticated, check for a present plugin
		if !authenticated {
			authenticated = CheckPluginAuth(username, password)
		}
	}

	if commonData.UseCache {
		authGranted := "false"
		if authenticated {
			authGranted = "true"
		}
		log.Debugf("setting auth cache for %s", username)
		SetAuthCache(username, password, authGranted)
	}

	return authenticated
}

//export AuthAclCheck
func AuthAclCheck(clientid, username, topic string, acc int) bool {

	aclCheck := false
	var cached = false
	var granted = false
	if commonData.UseCache {
		log.Debugf("checking acl cache for %s", username)
		cached, granted = CheckAclCache(username, topic, clientid, acc)
		if cached {
			log.Debugf("found in cache: %s", username)
			return granted
		}
	}

	//If prefixes are enabled, checkt if username has a valid prefix and use the correct backend if so.
	//Else, check all backends.
	if commonData.CheckPrefix {
		validPrefix, bename := CheckPrefix(username)
		if validPrefix {

			if bename == "plugin" {

				aclCheck = CheckPluginAcl(username, topic, clientid, acc)

			} else {

				var backend = commonData.Backends[bename]

				log.Debugf("Superuser check with backend %s", backend.GetName())
				if backend.GetSuperuser(username) {
					log.Debugf("superuser %s acl authenticated with backend %s", username, backend.GetName())
					aclCheck = true
				}

				//If not superuser, check acl.
				if !aclCheck {
					log.Debugf("Acl check with backend %s", backend.GetName())
					if backend.CheckAcl(username, topic, clientid, int32(acc)) {
						log.Debugf("user %s acl authenticated with backend %s", username, backend.GetName())
						aclCheck = true
					}
				}
			}

		} else {
			//If there's no valid prefix, check all backends.
			aclCheck = CheckBackendsAcl(username, topic, clientid, acc)
			//If acl hasn't passed, check for plugin.
			if !aclCheck {
				aclCheck = CheckPluginAcl(username, topic, clientid, acc)
			}
		}
	} else {
		aclCheck = CheckBackendsAcl(username, topic, clientid, acc)
		//If acl hasn't passed, check for plugin.
		if !aclCheck {
			aclCheck = CheckPluginAcl(username, topic, clientid, acc)
		}
	}

	if commonData.UseCache {
		authGranted := "false"
		if aclCheck {
			authGranted = "true"
		}
		log.Debugf("setting acl cache (granted = %s) for %s", authGranted, username)
		SetAclCache(username, topic, clientid, acc, authGranted)
	}

	log.Debugf("Acl is %t for user %s", aclCheck, username)

	return aclCheck
}

//export AuthPskKeyGet
func AuthPskKeyGet() bool {
	return true
}

//CheckAuthCache checks if the username/password pair is present in the cache. Return if it's present and, if so, if it was granted privileges.
func CheckAuthCache(username, password string) (bool, bool) {
	pair := b64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("auth%s%s", username, password)))
	val, err := commonData.RedisCache.Get(pair).Result()
	if err != nil {
		return false, false
	}
	//refresh expiration
	commonData.RedisCache.Expire(pair, time.Duration(commonData.AuthCacheSeconds)*time.Second)
	if val == "true" {
		return true, true
	}
	return true, false
}

//SetAuthCache sets a pair, granted option and expiration time.
func SetAuthCache(username, password string, granted string) error {
	pair := b64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("auth%s%s", username, password)))
	err := commonData.RedisCache.Set(pair, granted, time.Duration(commonData.AuthCacheSeconds)*time.Second).Err()
	if err != nil {
		return err
	}

	return nil
}

//CheckAclCache checks if the username/topic/clientid/acc mix is present in the cache. Return if it's present and, if so, if it was granted privileges.
func CheckAclCache(username, topic, clientid string, acc int) (bool, bool) {
	pair := b64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("acl%s%s%s%d", username, topic, clientid, acc)))
	val, err := commonData.RedisCache.Get(pair).Result()
	if err != nil {
		return false, false
	}
	//refresh expiration
	commonData.RedisCache.Expire(pair, time.Duration(commonData.AclCacheSeconds)*time.Second)
	if val == "true" {
		return true, true
	}
	return true, false
}

//SetAclCache sets a mix, granted option and expiration time.
func SetAclCache(username, topic, clientid string, acc int, granted string) error {
	pair := b64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("acl%s%s%s%d", username, topic, clientid, acc)))
	err := commonData.RedisCache.Set(pair, granted, time.Duration(commonData.AclCacheSeconds)*time.Second).Err()
	if err != nil {
		return err
	}

	return nil
}

//CheckPrefix checks if a username contains a valid prefix. If so, returns ok and the suitable backend name; else, !ok and empty string.
func CheckPrefix(username string) (bool, string) {
	if strings.Index(username, "_") > 0 {
		userPrefix := username[0:strings.Index(username, "_")]
		if prefix, ok := commonData.Prefixes[userPrefix]; ok {
			log.Debugf("Found prefix for user %s, using backend %s.", username, prefix)
			return true, prefix
		}
	}
	return false, ""
}

//CheckBackendsAuth checks for all backends if a username is authenticated and sets the authenticated param.
func CheckBackendsAuth(username, password string) bool {

	authenticated := false

	for _, bename := range backends {

		if bename == "plugin" {
			continue
		}

		var backend = commonData.Backends[bename]

		log.Debugf("checking user %s with backend %s", username, backend.GetName())

		if backend.GetUser(username, password) {
			authenticated = true
			log.Debugf("user %s authenticated with backend %s", username, backend.GetName())
			break
		}
	}

	return authenticated

}

//CheckBackendsAcl  checks for all backends if a username is superuser or has acl rights and sets the aclCheck param.
func CheckBackendsAcl(username, topic, clientid string, acc int) bool {
	//Check superusers first

	aclCheck := false

	for _, bename := range backends {

		if bename == "plugin" {
			continue
		}

		var backend = commonData.Backends[bename]

		log.Debugf("Superuser check with backend %s", backend.GetName())
		if backend.GetSuperuser(username) {
			log.Debugf("superuser %s acl authenticated with backend %s", username, backend.GetName())
			aclCheck = true
			break
		}
	}

	if !aclCheck {
		for _, bename := range backends {

			if bename == "plugin" {
				continue
			}

			var backend = commonData.Backends[bename]

			log.Debugf("Acl check with backend %s", backend.GetName())
			if backend.CheckAcl(username, topic, clientid, int32(acc)) {
				log.Debugf("user %s acl authenticated with backend %s", username, backend.GetName())
				aclCheck = true
				break
			}
		}
	}

	return aclCheck

}

//CheckPluginAuth checks that the plugin is not nil and returns the plugins auth response.
func CheckPluginAuth(username, password string) bool {
	if commonData.Plugin == nil {
		return false
	}
	return commonData.PGetUser(username, password)
}

//CheckPluginAcl checks that the plugin is not nil and returns the superuser/acl response.
func CheckPluginAcl(username, topic, clientid string, acc int) bool {
	if commonData.Plugin == nil {
		return false
	}
	//If superuser, authorize it.
	if commonData.PGetSuperuser(username) {
		return true
	}
	//Check against the plugin's check acl function.
	return commonData.PCheckAcl(username, topic, clientid, acc)
}

//export AuthPluginCleanup
func AuthPluginCleanup() {
	log.Info("Cleaning up plugin")
	//If cache is set, close cache connection.
	if commonData.RedisCache != nil {
		commonData.RedisCache.Close()
	}

	//Halt every registered backend.

	for _, v := range commonData.Backends {
		v.Halt()
	}

	if commonData.Plugin != nil {
		commonData.PHalt()
	}
}

func main() {}
